src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -28,9 +28,15 @@ private[concurrent] trait Promise[T] extends scala.concurrent.Promise[T] with sc
   import scala.concurrent.Future
   import scala.concurrent.impl.Promise.DefaultPromise
 
+  private[this] final def wrapFailure(t: Throwable): Throwable = {
+    if (NonFatal(t)) t
+    else if (t.isInstanceOf[InterruptedException]) new ExecutionException("Boxed InterruptedException", t)
```

Comment:
It seems 2.13 also [calls `Thread.currentThread.interrupt()`](https://github.com/scala/scala/blob/aa326fcc429d9a8e7f262eaab71f43a500d37b17/src/library/scala/concurrent/impl/Promise.scala#L445), should that be done here too?

https://github.com/scala/scala/pull/6610 says "Makes Futures correctly handle Thread.interrupt", I don't know what that entails.

Reply:
The new [extended example](https://docs.scala-lang.org/overviews/core/futures.html#exceptions) shows receiving the ExecutionException; on 2.12 it currently throws up the stack; with this fix, the world is right again. I just waffled on whether to re-interrupt in [this fix for vulpix](https://github.com/lampepfl/dotty/pull/17373/files) where `process.waitFor` is interrupted by `pool.shutdown`. It should interrupt (after retrying its work) in case subsequent work happens on the worker thread, even though one might assume that one is the last "client" of the thread. So I'd vote for interrupting the current (worker) thread.

I've forgotten the brouhaha over running side effects on the "current" thread; is there such an executor? But maybe it is irrelevant whether we know who created the current thread. If we caught the exception, we should set the interrupt status.

Reply:
I don't really understand that part and it's not obvious how to adapt it to the 2.12 design which is completely different. Since the new test suite passes with this version, it's obviously untested (and unspecified). Is the idea to always interrupt the current thread if a Promise was completed with a boxed InterruptedException?

Reply:
> Is the idea to always interrupt the current thread if a Promise was completed with a boxed InterruptedException?

That's how the 2.13 code reads.

Googlin' I find many places that say "if you catch an `InterruptedException`, you should call `Thread.currentThread.interrupt()`"
  - https://stackoverflow.com/questions/20934817/why-thread-currentthread-interrupt-be-called
  - https://stackoverflow.com/questions/4906799/why-invoke-thread-currentthread-interrupt-in-a-catch-interruptexception-block
  - https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html

summon[@viktorklang]?